### ISSUE TITLE
Add state machine composition

### ISSUE THREAD

**Issue by johnameyer** [johnameyer/cards-ts#113]
Currently our state machines exist as a flat structure

```
{
  [GameStates.START_GAME]: this.startGame,
  [GameStates.START_ROUND]: this.startRound,
  [GameStates.START_PASS]: this.startPass,
  [GameStates.WAIT_FOR_PASS]: this.waitForPass,
  [GameStates.HANDLE_PASS]: this.handlePass,
  [GameStates.START_FIRST_TRICK]: this.startFirstTrick,
  [GameStates.START_TRICK]: this.startTrick,
  [GameStates.START_PLAY]: this.startPlay,
  [GameStates.WAIT_FOR_PLAY]: this.waitForPlay,
  [GameStates.HANDLE_PLAY]: this.handlePlay,
  [GameStates.END_TRICK]: this.endTrick,
  [GameStates.END_ROUND]: this.endRound,
  [GameStates.END_GAME]: this.endGame,
}
```

To enable further composition and re-use, we must enable nesting of logic. For example, every startXyz, waitXyz, and handleXyz could all exist as a sub-machine (e.g. WaitStateTransitions), and loops might be their own sub-machine as well (though a loop is just a specialized general state). These constructs would be parameterized such that the specific behavior could be customized (or children states enabled). For the WaitStateTransitions, it would take in before and after handlers, and the event to call on which handlers. Likely the declaration and selection of the next state would remain in effect with the parent.

An open question is if the state controller should still be available for direct access in other controllers or in these handler functions.

Naturally, strong typing and proper serialization (meaning avoiding running handlers multiple times) is ideal. Likely the state variable would be represented by an array of states being pushed on.

Long term, we might want to consider supporting mermaid as an output for validation of flow.

**Comment by johnameyer** [comment]
Likely this can hook in with #91 wherein we would like to clearly establish which events are acceptable in response.

**Comment by johnameyer** [comment]
We should consider fixing up the "standard states" as a part of this

**Comment by johnameyer** [comment]
Potential euchre example:

```typescript
const stateMachine = loop({
  id: 'round',
  beforeEach: controllers => { /* clean up state */ },
  afterAll: controllers => { /* update scores and send messages */ },
  while: controllers => /* no player has yet won */,
  do: sequence([
    compose([
      handle({ // specialized variant - can be loop or single state
        id: 'order-up', // these can probably be left off if there is uniqueness handling in place
        beforeAll: controllers => { /* message about flipped card */ },
        // TODO ponder about semantics between outer beforeEach and inner beforeAll
        afterEach: controllers => { /* message (incl. going along) */ }
        // TODO iterate players once starting from dealer, breaking if any player selects a card 
        handler: 'orderUp',
        args: [], // leave off if empty, otherwise controllers => {}
      }),
      handle({
        handler: 'dealerDiscard',
        // TODO identify player
        after: controllers => { /* take the discard and give new card */ }, // plays into #116 
      }),
      handle({
        // TODO iterate players once starting from dealer, breaking if any player selects a card
        // TODO with compose, how do we designate 'else' state - no player picks
        // perhaps instead we support promise-like functions like machine.andThen(), catch()
        // but then there is a question of divergence i.e. A -> C, D and B -> C, D
        handler: 'nameTrump',
        afterAll: controllers => { /* message (incl. going along) */ }
      }),
    ]),
    loop({
      id: 'trick',
      beforeEach: controllers => { /* send messages, reset states */ },
      afterEach: controllers => { /* message, update leader, update count */ },
      while: controllers => /* players have cards */,
      do: handle({
        // TODO iterate players once, starting from leader, if playing
        handler: 'turn',
        after: controllers => { /* send message (setting state currently handled by event-handler merge */ }
      })
    }),
  ]),
})
```

Some other questions:

* Is there some higher-level construct we should be exposing around asking players?
* Is there some mechanism to enforce unique ids at build/compile time? If users can still manually build the structures how does that work? If there is some abstraction around the ids, how can they be referred to by submachines?
* Where do we fit in start / end game handlers

**Comment by johnameyer** [comment]
As a part of the build / test, we should consider generating a state diagram (mermaid?).

We should try to have the transitions named as a result
